DEBIAN/svnnotify.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2013 Markus <mog@equinox>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH SVNNOTIFY SECTION "July 29, 2013"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
svnnotify \- program to do something
.SH SYNOPSIS
.B svnnotify
.RI [ options ] " files" ...
.br
.B bar
.RI [ options ] " files" ...
.SH DESCRIPTION
This manual page documents briefly the
.B svnnotify
and
.B bar
commands.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBsvnnotify\fP is a program that...
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.SH SEE ALSO
.BR bar (1),
.BR baz (1).
.br
The programs are documented fully by
.IR "The Rise and Fall of a Fooish Bar" ,
available via the Info system.


debian/changelog
svnnotify (0.3-1ubuntu1) UNRELEASED; urgency=low

  * added desktop icon

 -- mog33 <ctrl.alt.del@gmx.net>  Mon, 05 Aug 2013 21:59:51 +0200

svnnotify (0.3-1) unstable; urgency=low

  * Minor bugfixes
  * correct path as working directory
  * using debhelper for creating debian package

 -- mog33 <ctrl.alt.del@gmx.net>  Tue, 30 Jul 2013 08:58:31 +0200

svnnotify (0.2-1) unstable; urgency=low

  * Initial release 

 -- mog33 <ctrl.alt.del@gmx.net>  Mon, 29 Jul 2013 18:52:36 +0200


svnnotify.py
#!/usr/bin/python
#
# Script to notify the user for changes in a subversion repository.
#
# Depends on a configuration file with one or more of the following entries:
#
#   [ServerName]
#   server=SVN_REPO_TO_MONITOR
#   user=YOUR_SVN_USERNAME
#   pass=YOUR_SVN_PASSWORD
#

import pynotify
import datetime, time, os, logging
from pysvn import Client, Revision, opt_revision_kind, ClientError
from appdirs import user_data_dir


class SvnNotifier():

    def __init__(self, name, svn_root, svn_username, svn_password, parser, config_file):
        self.name = name
        self.svn_root = svn_root
        self.svn_username = svn_username
        self.svn_password = svn_password
        self.client = Client()
        self.client.callback_get_login = self.credentials
        self.parser = parser
        self.config_file = config_file

    def notify(self, author, date, revision, message, paths):
        """Display the changed paths using libnotify"""
        title_string = 'New commit #%s to repository %s' % (revision, self.name)
        message_string = "<p>[%s] %s</p> <p><i>%s</i>" % \
                        (date.strftime("%d.%m %H:%M"), author, message)
        message_string += "<ul>"
        for p in paths[:5]:
            if len(p)>50:
                p = "...%s" % p[-50:]
            message_string += "<li>%s</li>" % p
        if len(paths)>6:
            message_string += "<li>...</li>"
        message_string += "</ul></p>"
        logging.debug("Open pynotify.Notification: %s | %s" % (title_string,message_string))
        pynotify.Notification(title_string, message_string, "view-refresh").show()

    def credentials(self, realm, username, may_save):
        """Return the default login credentials"""
        return True, self.svn_username, self.svn_password, False

    def discover_changes(self):
        """
        Find out the changes occured since the last time this method is ran
        """
        logging.debug("Discover Changes in %s" % self.name)
        self.parser.read(self.config_file)
        if self.parser.has_option(self.name, "last_revision"):
            last_revision = self.parser.get(self.name, "last_revision")
        else:
            last_revision = 0
        log = self.client.log(
            self.svn_root,
            discover_changed_paths=True,
            revision_end=Revision(opt_revision_kind.number, last_revision)
            )
        log = log[:-1]   # Ignore last revision
        if len(log) > 0:
            logging.info("%s new commits for %s" % (len(log), self.name))
            # update last revision in config file
            last_revision = log[0].revision.number
            self.parser.set(self.name, "last_revision", last_revision)
            self.parser.write(open(self.config_file, 'w'))
            log.reverse()
            if len(log)>5: # Show only last 10 commits
                pynotify.Notification("Even more commits for repository %s" % self.name, "There are %s more new commits in the repository" % (len(log)-5), "view-refresh").show()
                log = log [-5:]
            for entry in log:
                author = entry.get('author')
                rev = entry.revision.number
                message = entry.message
                date = datetime.datetime.fromtimestamp(entry.date)
                paths = []
                for change in entry.changed_paths:
                    if change.path not in paths:
                        paths.append(change.path)
                self.notify(author, date, rev, message, paths)
        else:
            logging.info("No new commits for %s" % self.name)


class Configuration():

    def __init__(self, config_file):
        if config_file:
            self.config_file = config_file
        else:
            self.config_file = self.get_config_file()
        logging.info("SvnNotify started with configuration file %s" % config_file)

    def get_config_file(self):
        cfg = os.path.join(os.getcwd(),"svnnotify.cfg")
        if os.path.isfile(cfg):
            return cfg
        cfg = os.path.join(user_data_dir("svnnotify"),"svnnotify.cfg")
        if os.path.isfile(cfg):
            return cfg
        else:
            if not os.path.exists(user_data_dir("svnnotify")):
                os.makedirs(user_data_dir("svnnotify"))
            f = open(cfg, "w")
            f.write("""#[Repo1Name]
#server = https://server1.com/svn/repo
#user = username
#pass = password

#[Repo2Name]
#server = svn+ssh://username@server2.com/svn/repo""") 
            print("Configuration file for svnnotify created: %s" % cfg)
            print("Modify it to your needs before restarting")
            exit()


    def create_repos_from_svn_config(self):
        from ConfigParser import ConfigParser
        parser = ConfigParser()
        parser.read(self.config_file)
        repos = []
        for section in parser.sections():
            try:
                server = parser.get(section, 'server')
            except BaseException as e:
                logging.critical("Error while parsing config file %s\n%s" % (self.config_file,e))
                exit()
            if parser.has_option(section, 'user'):
                user = parser.get(section, 'user')
            else:
                user = None
            if parser.has_option(section, 'pass'):
                passw = parser.get(section, 'pass')
            else:
                passw = None
            repos.append(SvnNotifier(section, server, user, passw, parser, self.config_file))
            logging.info('Monitoring SVN repository %s (%s)' % (section, server))
        if repos:
            return repos
        else:
            print("No sections in configuration file found. Aborting")
            exit()


def svn_notify(config_file, interval):
    path = os.path.dirname(__file__)
    logging.basicConfig(
        level = logging.INFO,
        format = "%(asctime)s %(levelname)s: %(message)s",
        datefmt = "%d.%m.%Y %H:%M:%S",
        #filename= os.path.join(path, 'svnnotify.log')
        )
    conf = Configuration(config_file)
    repos = conf.create_repos_from_svn_config()
    pynotify.init("SVN Monitor")
    pynotify.Notification("SvnNotify started", "SvnNotify started with configuration file %s monitoring %i repositories." % (conf.config_file, len(repos)), "view-refresh").show()
    while True:
        for repo in repos:
            try:
                repo.discover_changes()
            except ClientError as e:
                logging.error("Repo %s not accessible\n%s" % (repo.name,e))
        time.sleep(interval)


if __name__ == '__main__':
    from argparse import ArgumentParser, FileType
    parser = ArgumentParser(description="Notifies about changes in SVN repositories")
    parser.add_argument("-c","--config", help="path to configuration file (default svnnotify.cfg in working directory or in %s)" % user_data_dir("svnnotify"))
    parser.add_argument("-v","--version", help="shows version of svnnotify", action='version', version='%(prog)s 0.3')
    parser.add_argument("-i","--interval", help="update interval in seconds (default 5 minutes)", type=int, default=60*5)
    args = parser.parse_args()
    svn_notify(args.config, args.interval)
